### ISSUE TITLE
Track recording consumes too much battery

### ISSUE THREAD

**Issue by ioctl-user** [organicmaps/organicmaps#9375]
For now track recording is very energy intensive. Here is Battery historian tool picture with Organic Maps track record of my phone with Qualcomm Snapdragon 865 CPU in Flight mode.

![orgmaps](https://github.com/user-attachments/assets/7d894138-6407-4047-9a8e-2f991288a7bf)

Power consumption is about 4.2% per hour. I cannot export GPX file, but may assume, position recorded each 15 seconds. At that, as you can see from the picture, app wakes CPU each one second.

Here is the same picture of uLogger app from the F-Droid with 15-second interval record set:

![ulogger](https://github.com/user-attachments/assets/9c2426fd-caa4-47b9-9fc9-3bffe17aa681)

Power consumption is about 0.85% per hour. App wakes for a few short periods each 15 seconds, which is much better.

Please consider optimize track recording power consumption.

**Comment by ioctl-user** [comment]
Related issues:

https://github.com/organicmaps/organicmaps/issues/8927
https://github.com/organicmaps/organicmaps/issues/318

**Comment by Rob-from-VI** [comment]
As to the battery consumtion itself, I have not so far found it to be excessive at all.   I will run some tests and see if I can establish a baseline.

**Comment by Rob-from-VI** [comment]
I recorded a track for 15 minutes with phone fully active, but screen off.

Accubattery app reported that Organic Maps beta had used 20 mAh, or .5% on my Samsung S20FE

**Comment by Rob-from-VI** [comment]
Today I recorded a track for 2.5 hours mostly walking at good speed, most of the time screen off in my pocket, occasional check that it was recording, and again according to Accubattery, Organic maps beta consumed only 38.4 mAh.

**Comment by ioctl-user** [comment]
Additional info:

My phone itself in the Flight Mode with screen of consumes 0.35% of batter according to the [Batter Historian](https://developer.android.com/topic/performance/power/setup-battery-historian).

So, default Organic Maps recording itself consumes 4.2% - 0.35% which gives 3.85% per hour.
Well optimized [μlogger ](https://f-droid.org/en/packages/net.fabiszewski.ulogger/) consumes itself 0.5% per hour.

It's almost an 8-times difference, which is very valuable while recording long-time tracks.

**Comment by Ishatix** [comment]
I did a rough first test of track recording while out for a walk the other day. It wasn't in any way scientific, but it used up ~10% of my battery per hour on average which is significantly more than needed for the app I usually use to record GPS traces for tagging photos with (OI. Track).

**Comment by Rob-from-VI** [comment]
Interesting - mine used about 10% overall in my 2.5 hour hike, but I had lots of other apps in the background, and I had poor connectivity so it spent a lot of energy looking for service.

When I used OSMand in the past, the battery consumption while recording was about the same as Organic is now.

**Comment by ioctl-user** [comment]
> When I used OSMand in the past, the battery consumption while recording was about the same as Organic is now.

The point is OSMand is not optimal also. I have created bug a few years ago about this problem, but nothing happen.

> I will try and do another test with all other apps turned off, and in flight mode.

Consider using Battery Historian. Google mark this tool as obsolete, but it's steel working fine in the official Docker image. It can show each CPU wake app event, battery consumption with high precision and a lot of another parameters.

**Comment by Ishatix** [comment]
My first test of track recording above was just on a short few hours walk. I carried out a second test today on a longer cycle ride, setting out with 90% battery charge and in flight mode with no other apps in use or left in memory / open in the background.

For the first ~1.5 hours I had OM open (with the desired route just set up and shown on the standard top down map display, but not following along in the "sat nav" style mode). The screen was mostly switched off, and only switched on for a few seconds 3 or 4 times in total over that leg to check a few turnings. At the end of that leg my battery had already dropped to 64%. If I had not been recording the track, I would have expected it to drop only 2 or 3 percent over that period.

For the remaining ~4 to 4.5 hours of the cycle, I did not need the map at all, so I exited/closed down OM (leaving track recording running in the background), and barely touched my phone again until I got home. At the end I still had 51% battery left. So for that remaining period the battery usage was still a little on the high side, but far more reasonable. 

So while there may still be room for improvement when recording in background, it appears that there is a particularly major battery drain occurring when recording with the app open. ...Could this even be an issue with some other part of the app that is normally not active when the screen is switched off, but is kept active by the track recording feature?? In any case, it would be interesting if others could test out these two scenarios to confirm if it is a repeatable correlation, and not some kind of intermittent behaviour. I did notice on my first test that the battery seemed to drop more quickly at the start of my walk, which perhaps might have the same explanation behind it.

This was all using 2024.09.09-7-FDroid running on Android 7.

P.S. At the end of this track, I selected "Save" after stopping it, and the track simply disappeared. When looking in my "My Places" collection of bookmarks, I could still see my first test at the top of the list from last week (and displayed on the map), but no sign of the new recording today. I thought it had been lost for good. But on double checking through everything more carefully now, I have finally located it. I had imported a track into OM a few weeks ago just to test out that functionality. And imported tracks are shown as their own separate list entry. Upon opening that old imported track, I have found that today's track has been saved under it into that list for some reason. That imported track is the first displayed entry at the top of the bookmarks list, just above "My Places", but I thought new tracks were supposed to get saved into the last used list, and since my previous test was saved into "My Places", and I didn't do anything differently this time, it is a bit difficult to make any sense of. Anyway, if the default track record save behaviour is likely to be changed soon in any case, I guess there's little value in opening a bug report for this?

**Comment by kirylkaveryn** [comment]
@ioctl-user can you please make a test while running multiple track recording apps **at the same time**?

I've made a test on iOS and the OM's track recording consumed **8x less** power than GuruMaps and 2x less than OsmAnd on 2.5h recording. Toe OM ate only 2% till 2.5 hours. The apps was in the background almost all the time.

It's important to understand: is it the Andoid-specific issue or is there some bug in the core implementation?
Additional tests on iOS will be helpful!

**Comment by ioctl-user** [comment]
@kirylkaveryn , my Lineage OS version has quite limited battery usage statistics UI. I cannot get battery consumption for a given period. And even general statistics don't seem adequate.

Anyway, running app simultaneously is not reliable method check each app own power consumption, because Android can aggregate planned wakelocks.

I think even with not well calibrated battery, running apps one by one with medium battery charged will give more precise results.

I'd like to point out again that OM wakes phone each seconds while recording position once about 15 seconds. Seems, this is cause of the problem.

**Comment by pastk** [comment]
> I'd like to point out again that OM wakes phone each seconds while recording position once about 15 seconds. Seems, this is cause of the problem.

OM is supposed to record points with a 1 second interval (@kavikhalique is that right?)
Where did you get the 15 seconds figure from?

OM might actually write points less frequently if adjacent points are too close too each other (filtering out the noise and tiny movements).

**Comment by pastk** [comment]
There is an idea that a lot of power is consumed for writing to persistent storage.

If this is the case indeed, then adding a short caching buffer  (e.g. 10-15 seconds) could reduce battery consumption.

There is a possibility to lose the last 10-15 seconds of the track in case of a crash or app getting killed by the system, etc. But 10-15 seconds seems like a very short time compared to e.g. time it takes to notice the outage and restart the app (or the phone) to resume recording.

**Comment by kavikhalique** [comment]
> OM is supposed to record points with a 1 second interval (@kavikhalique is that right?)
> 

Yes, Android app of OM sends co-ordinates every second to core then it filters out points according to accuracy and distance with previously saved point.

**Comment by Rob-from-VI** [comment]
I actually think it is also based on distance somehow as well.   

I have attempted to make it record one final point at my exact destination, and while walking around in small circles for some time (more than 15 seconds) no aditional point gets recorded, but if I walk more than about 13 m beyond the last point, then and only then it records a new point.

It would sure be nice if somebody that knows for sure how this presently works would chime in here and tell us.  

I would actually like it to record one last point when I press stop, and before I save the track

**Comment by Rob-from-VI** [comment]
For those who are experiencing high battery consumption, is your screen coming on while recording the track?   Mine stays off unless i open it to look at what has been recorded.

**Comment by ioctl-user** [comment]
> For those who are experiencing high battery consumption, is your screen coming on while recording the track?

Sure it's off. And phone is in the flight mode.

**Comment by Ishatix** [comment]
> For those who are experiencing high battery consumption, is your screen coming on while recording the track? Mine stays off unless i open it to look at what has been recorded.

No, but over the past 6 months or more I'm starting to think (that at least for me on Android 7), the issue might be more simply with high battery usage in general from the app staying active in the background, as opposed to (just) when recording a track. I'm not sure that it is always the case, but at least on occasion I have noticed high battery drain from the app just from leaving it open (i.e. just switching off the phone screen without closing the app). I'm only more likely to do that in cases where I've setup a route to follow, or recording a track. So it's difficult to be sure how much either of these features might be contributing, but I certainly recall experiencing it in cases when I wasn't recording a track.

I have also noticed that the app can be extremely power hungry when in active usage (with rendering the map while moving it around a lot?). E.g. I was doing some research with it recently and it was noticeably draining the battery even while the phone was plugged into a charger. I guess this should be considered a separate issue though, albeit with presumably some potential overlap.

**Comment by Ishatix** [comment]
Hi all,

I spent 3 months earlier this year hiking around Malaysia and Australia using Organic Maps extensively every day across two different phones (Both Samsung S7 Active / SAMSUNG-SM-G891A, one running Android 7 and the other Android 8. From numerous experiences along the way, I am now quite certain that there is a sporadic/intermittent excessive battery usage bug still present in OM, that is more related to just keeping the map view active (with screen switched off), than specifically anything to do with track recording.

My usual manner of usage is to have a route planned out on the map for at least just the next section of my walk that I have decided upon, and leave that marked in place on the usual top-down / compass-point orientated map (i.e. I will never hit the "Start" button as I do not want the isometric POV that imposes). I believe it is (only?) when I leave that setup route in place, and just directly switch off the phone display from there, that very occasionally the app will still be left doing something in the background that very quickly eats through my battery (and the app gets reported as being a major battery drain by the Android system settings battery monitor). 

As this happens only occasionally/intermittently, it is difficult to test to be sure, but I'm almost certain it never happens if I exit out of the app with the phone back button before switching off the display (and thus lose my planned route, which can be quite a pain if I've added in quite a few added stops to augment it along the way). I think it might also not happen if I simply minimise the app with the home button, before switching off the display (and thus can keep the route in place, and so this became my preferred approach for avoiding the risk).

Not sure how else I could investigate further... I see there is an option in the app to report a bug that appears to export logs. Would it be worth trying to capture and submit those the next time I can get the issue to manifest?

**Comment by biodranik** [comment]
Hi and thanks for the detailed report!

- Do you have active track recording enabled?
- Do you have any per-app battery information displayed in your system that can prove Organic Map (and not some other background app) excessive battery usage?
- There could be a bug related to background upload of OpenStreetMap.org edits. Did you do any?
- How old are the batteries on your devices?

It would be great to confirm it on newer Android versions.

**Comment by Ishatix** [comment]
Thanks for interest! In reply:


> * Do you have active track recording enabled?

No. Hence it is not the same as any issues mentioned above with regard to the baseline battery consumption of that feature. I noticed issues with this myself, but then came to the conclusion that it was not the track recording feature _per se_ that is eating the battery (at least in my case/testing), but (or additionally) an issue related to how much the app/map remains active (after switching off the display) as is more likely the case when recording a track.
Probably I should do some more testing with track recording too at some point, to see if there is more variability in usage with that. 


> * Do you have any per-app battery information displayed in your system that can prove Organic Map (and not some other background app) excessive battery usage?

Yes. In the cases where it happens, the Android battery management system clearly flags Organic Maps as the app that is causing the excessive drainage (and in many cases it is the only app I am using over the course of the day on that device, with any other app requirements being contained to a separate device). And restarting the app resolves the issue.


> * There could be a bug related to background upload of OpenStreetMap.org edits. Did you do any?

No, I never have (to my great shame!).


> * How old are the batteries on your devices?

Very old. But this is irrelevant as it is about relative usage. Most of the time it is not an issue at all, just very occasionally something happens to trigger it. But I have never found a way to trigger it intentionally to repeat it for demonstration purposes unfortunately  :-/

**Comment by biodranik** [comment]
@Ishatix you can enable logs recording in OM settings next time, and if the issue is reproduced, send us logs via Report a bug, and tell at what time did you observe extensive battery usage.

**Comment by Ishatix** [comment]
> [@Ishatix](https://github.com/Ishatix) you can enable logs recording in OM settings next time, and if the issue is reproduced, send us logs via Report a bug, and tell at what time did you observe extensive battery usage.

Good idea thanks. Is enabling this logging likely to add much additional battery drain? If not, I'm happy to just leave it enabled permanently.

**Comment by biodranik** [comment]
Would be great to measure it too :)

In general, always writing logs is not recommended.
